### PR TITLE
Deduplicate ChangesetEvents per Changeset before DB insert

### LINKED ISSUE #10264: GitHub review_requested ChangesetEvents break syncer if the requested team has been deleted
The syncing process on my machine currently fails to execute this call:

https://github.com/sourcegraph/sourcegraph/blob/96b8fe4c463447da3575eddfb7f9cb067348bec1/enterprise/internal/campaigns/syncer.go#L576

The logged error:

```
ERROR Syncing changeset, err: pq: ON CONFLICT DO UPDATE command cannot affect row a second time
```

Looking in my PostgreSQL logs revealed that the `UpsertChangesetEvents` query contains duplicate entries for changeset events:

<details>

```json
[
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:commit",
        "key": "e7a54e12ed1e3f0ad35fc0ba0be9b91049b6a5c1",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Commit": {
                "OID": "e7a54e12ed1e3f0ad35fc0ba0be9b91049b6a5c1",
                "Message": "Make campaign.description nullabke",
                "MessageHeadline": "Make campaign.description nullabke",
                "URL": "https://github.com/sourcegraph/sourcegraph/commit/e7a54e12ed1e3f0ad35fc0ba0be9b91049b6a5c1",
                "Committer": {
                    "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                    "Email": "erikseliger@me.com",
                    "Name": "Erik Seliger",
                    "User": {
                        "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                        "Login": "eseliger",
                        "URL": "https://github.com/eseliger"
                    }
                },
                "CommittedDate": "2020-04-09T14:42:34Z",
                "PushedDate": "2020-04-09T14:42:44Z"
            }
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:labeled",
        "key": "MDU6TGFiZWwxMDMzNjAzMTU2:add:1586443388000000000",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Actor": {
                "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                "Login": "eseliger",
                "URL": "https://github.com/eseliger"
            },
            "Label": {
                "ID": "MDU6TGFiZWwxMDMzNjAzMTU2",
                "Color": "795548",
                "Description": "Technical debt.",
                "Name": "debt"
            },
            "CreatedAt": "2020-04-09T14:43:08Z",
            "Removed": false
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:labeled",
        "key": "MDU6TGFiZWwxNTMyNTU0NDkz:add:1586443388000000000",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Actor": {
                "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                "Login": "eseliger",
                "URL": "https://github.com/eseliger"
            },
            "Label": {
                "ID": "MDU6TGFiZWwxNTMyNTU0NDkz",
                "Color": "ef820a",
                "Description": "",
                "Name": "campaigns"
            },
            "CreatedAt": "2020-04-09T14:43:08Z",
            "Removed": false
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:review_requested",
        "key": "eseliger:ryanslade:1586443388000000000",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Actor": {
                "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                "Login": "eseliger",
                "URL": "https://github.com/eseliger"
            },
            "RequestedReviewer": {
                "AvatarURL": "https://avatars3.githubusercontent.com/u/25610?u=416aa7bd7c7a97c714ea0a503c90a0e7e21c5e56\u0026v=4",
                "Login": "ryanslade",
                "URL": "https://github.com/ryanslade"
            },
            "RequestedTeam": {
                "Name": "",
                "URL": ""
            },
            "CreatedAt": "2020-04-09T14:43:08Z"
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:review_requested",
        "key": "eseliger:mrnugget:1586443388000000000",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Actor": {
                "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                "Login": "eseliger",
                "URL": "https://github.com/eseliger"
            },
            "RequestedReviewer": {
                "AvatarURL": "https://avatars1.githubusercontent.com/u/1185253?u=35f048c505007991433b46c9c0616ccbcfbd4bff\u0026v=4",
                "Login": "mrnugget",
                "URL": "https://github.com/mrnugget"
            },
            "RequestedTeam": {
                "Name": "",
                "URL": ""
            },
            "CreatedAt": "2020-04-09T14:43:08Z"
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:review_requested",
        "key": "eseliger:slimsag:1586443388000000000",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Actor": {
                "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                "Login": "eseliger",
                "URL": "https://github.com/eseliger"
            },
            "RequestedReviewer": {
                "AvatarURL": "https://avatars1.githubusercontent.com/u/3173176?u=4d116636ebe647649a2157490c8d4fc21ad1fd16\u0026v=4",
                "Login": "slimsag",
                "URL": "https://github.com/slimsag"
            },
            "RequestedTeam": {
                "Name": "",
                "URL": ""
            },
            "CreatedAt": "2020-04-09T14:43:08Z"
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:review_requested",
        "key": "eseliger::1586443388000000000",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Actor": {
                "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                "Login": "eseliger",
                "URL": "https://github.com/eseliger"
            },
            "RequestedReviewer": {
                "AvatarURL": "",
                "Login": "",
                "URL": ""
            },
            "RequestedTeam": {
                "Name": "",
                "URL": ""
            },
            "CreatedAt": "2020-04-09T14:43:08Z"
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:review_requested",
        "key": "eseliger::1586443388000000000",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Actor": {
                "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                "Login": "eseliger",
                "URL": "https://github.com/eseliger"
            },
            "RequestedReviewer": {
                "AvatarURL": "",
                "Login": "",
                "URL": ""
            },
            "RequestedTeam": {
                "Name": "",
                "URL": ""
            },
            "CreatedAt": "2020-04-09T14:43:08Z"
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:review_requested",
        "key": "eseliger:Core Services:1586443388000000000",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Actor": {
                "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                "Login": "eseliger",
                "URL": "https://github.com/eseliger"
            },
            "RequestedReviewer": {
                "AvatarURL": "",
                "Login": "",
                "URL": ""
            },
            "RequestedTeam": {
                "Name": "Core Services",
                "URL": "https://github.com/orgs/sourcegraph/teams/core-services"
            },
            "CreatedAt": "2020-04-09T14:43:08Z"
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:review_requested",
        "key": "eseliger:web:1586443388000000000",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Actor": {
                "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                "Login": "eseliger",
                "URL": "https://github.com/eseliger"
            },
            "RequestedReviewer": {
                "AvatarURL": "",
                "Login": "",
                "URL": ""
            },
            "RequestedTeam": {
                "Name": "web",
                "URL": "https://github.com/orgs/sourcegraph/teams/web"
            },
            "CreatedAt": "2020-04-09T14:43:08Z"
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:reviewed",
        "key": "391019934",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "DatabaseID": 391019934,
            "Author": {
                "AvatarURL": "https://avatars3.githubusercontent.com/u/2946214?u=0cdcf6f1d7de3a06e649fd5f1c40763dcc031e98\u0026v=4",
                "Login": "unknwon",
                "URL": "https://github.com/unknwon"
            },
            "AuthorAssociation": "MEMBER",
            "Body": "",
            "State": "APPROVED",
            "URL": "https://github.com/sourcegraph/sourcegraph/pull/9702#pullrequestreview-391019934",
            "CreatedAt": "2020-04-09T18:10:33Z",
            "UpdatedAt": "2020-04-09T18:10:33Z",
            "Commit": {
                "OID": "e7a54e12ed1e3f0ad35fc0ba0be9b91049b6a5c1",
                "Message": "Make campaign.description nullabke",
                "MessageHeadline": "Make campaign.description nullabke",
                "URL": "https://github.com/sourcegraph/sourcegraph/commit/e7a54e12ed1e3f0ad35fc0ba0be9b91049b6a5c1",
                "Committer": {
                    "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                    "Email": "erikseliger@me.com",
                    "Name": "Erik Seliger",
                    "User": {
                        "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                        "Login": "eseliger",
                        "URL": "https://github.com/eseliger"
                    }
                },
                "CommittedDate": "2020-04-09T14:42:34Z",
                "PushedDate": "2020-04-09T14:42:44Z"
            },
            "IncludesCreatedEdit": false
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:renamed",
        "key": "mrnugget:1586762219000000000",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Actor": {
                "AvatarURL": "https://avatars1.githubusercontent.com/u/1185253?u=35f048c505007991433b46c9c0616ccbcfbd4bff\u0026v=4",
                "Login": "mrnugget",
                "URL": "https://github.com/mrnugget"
            },
            "PreviousTitle": "Make campaign.description nullabke",
            "CurrentTitle": "Make campaign.description nullable",
            "CreatedAt": "2020-04-13T07:16:59Z"
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:reviewed",
        "key": "391996087",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "DatabaseID": 391996087,
            "Author": {
                "AvatarURL": "https://avatars1.githubusercontent.com/u/1185253?u=35f048c505007991433b46c9c0616ccbcfbd4bff\u0026v=4",
                "Login": "mrnugget",
                "URL": "https://github.com/mrnugget"
            },
            "AuthorAssociation": "MEMBER",
            "Body": "Approve, but I think we should be consistent and also make it nullable in the database :)",
            "State": "APPROVED",
            "URL": "https://github.com/sourcegraph/sourcegraph/pull/9702#pullrequestreview-391996087",
            "CreatedAt": "2020-04-13T07:18:21Z",
            "UpdatedAt": "2020-04-13T07:18:21Z",
            "Commit": {
                "OID": "e7a54e12ed1e3f0ad35fc0ba0be9b91049b6a5c1",
                "Message": "Make campaign.description nullabke",
                "MessageHeadline": "Make campaign.description nullabke",
                "URL": "https://github.com/sourcegraph/sourcegraph/commit/e7a54e12ed1e3f0ad35fc0ba0be9b91049b6a5c1",
                "Committer": {
                    "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                    "Email": "erikseliger@me.com",
                    "Name": "Erik Seliger",
                    "User": {
                        "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                        "Login": "eseliger",
                        "URL": "https://github.com/eseliger"
                    }
                },
                "CommittedDate": "2020-04-09T14:42:34Z",
                "PushedDate": "2020-04-09T14:42:44Z"
            },
            "IncludesCreatedEdit": false
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:commit",
        "key": "11f72f45611f77d6c0cbcb6fbe74266a74eab0a4",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Commit": {
                "OID": "11f72f45611f77d6c0cbcb6fbe74266a74eab0a4",
                "Message": "Merge branch ''master'' into campaigns/nullable-description",
                "MessageHeadline": "Merge branch ''master'' into campaigns/nullable-description",
                "URL": "https://github.com/sourcegraph/sourcegraph/commit/11f72f45611f77d6c0cbcb6fbe74266a74eab0a4",
                "Committer": {
                    "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                    "Email": "erikseliger@me.com",
                    "Name": "Erik Seliger",
                    "User": {
                        "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                        "Login": "eseliger",
                        "URL": "https://github.com/eseliger"
                    }
                },
                "CommittedDate": "2020-04-13T12:09:23Z",
                "PushedDate": "0001-01-01T00:00:00Z"
            }
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:commit",
        "key": "b552dcdc77abec8b55559595187f8df7e31127cb",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Commit": {
                "OID": "b552dcdc77abec8b55559595187f8df7e31127cb",
                "Message": "Make description column nullable",
                "MessageHeadline": "Make description column nullable",
                "URL": "https://github.com/sourcegraph/sourcegraph/commit/b552dcdc77abec8b55559595187f8df7e31127cb",
                "Committer": {
                    "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                    "Email": "erikseliger@me.com",
                    "Name": "Erik Seliger",
                    "User": {
                        "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                        "Login": "eseliger",
                        "URL": "https://github.com/eseliger"
                    }
                },
                "CommittedDate": "2020-04-13T12:20:47Z",
                "PushedDate": "2020-04-13T12:21:39Z"
            }
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:commented",
        "key": "612876476",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "DatabaseID": 612876476,
            "Author": {
                "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                "Login": "eseliger",
                "URL": "https://github.com/eseliger"
            },
            "Editor": null,
            "AuthorAssociation": "MEMBER",
            "Body": "I added it, PTAL if I changed all required pieces :) ",
            "URL": "https://github.com/sourcegraph/sourcegraph/pull/9702#issuecomment-612876476",
            "CreatedAt": "2020-04-13T12:21:46Z",
            "UpdatedAt": "2020-04-13T12:21:46Z",
            "IncludesCreatedEdit": false
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:reviewed",
        "key": "392113697",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "DatabaseID": 392113697,
            "Author": {
                "AvatarURL": "https://avatars1.githubusercontent.com/u/1185253?u=35f048c505007991433b46c9c0616ccbcfbd4bff\u0026v=4",
                "Login": "mrnugget",
                "URL": "https://github.com/mrnugget"
            },
            "AuthorAssociation": "MEMBER",
            "Body": "Looks good to me!",
            "State": "APPROVED",
            "URL": "https://github.com/sourcegraph/sourcegraph/pull/9702#pullrequestreview-392113697",
            "CreatedAt": "2020-04-13T12:25:08Z",
            "UpdatedAt": "2020-04-13T12:25:08Z",
            "Commit": {
                "OID": "b552dcdc77abec8b55559595187f8df7e31127cb",
                "Message": "Make description column nullable",
                "MessageHeadline": "Make description column nullable",
                "URL": "https://github.com/sourcegraph/sourcegraph/commit/b552dcdc77abec8b55559595187f8df7e31127cb",
                "Committer": {
                    "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                    "Email": "erikseliger@me.com",
                    "Name": "Erik Seliger",
                    "User": {
                        "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                        "Login": "eseliger",
                        "URL": "https://github.com/eseliger"
                    }
                },
                "CommittedDate": "2020-04-13T12:20:47Z",
                "PushedDate": "2020-04-13T12:21:39Z"
            },
            "IncludesCreatedEdit": false
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:commit",
        "key": "68c7e735576b7ecf7e062877319e2b115c752ba8",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Commit": {
                "OID": "68c7e735576b7ecf7e062877319e2b115c752ba8",
                "Message": "Update schema",
                "MessageHeadline": "Update schema",
                "URL": "https://github.com/sourcegraph/sourcegraph/commit/68c7e735576b7ecf7e062877319e2b115c752ba8",
                "Committer": {
                    "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                    "Email": "erikseliger@me.com",
                    "Name": "Erik Seliger",
                    "User": {
                        "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                        "Login": "eseliger",
                        "URL": "https://github.com/eseliger"
                    }
                },
                "CommittedDate": "2020-04-13T12:41:20Z",
                "PushedDate": "2020-04-13T12:41:28Z"
            }
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:commented",
        "key": "612883690",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "DatabaseID": 612883690,
            "Author": {
                "AvatarURL": "https://avatars2.githubusercontent.com/in/254?v=4",
                "Login": "codecov",
                "URL": "https://github.com/apps/codecov"
            },
            "Editor": null,
            "AuthorAssociation": "NONE",
            "Body": "# [Codecov](https://codecov.io/gh/sourcegraph/sourcegraph/pull/9702?src=pr\u0026el=h1) Report\n\u003e Merging [#9702](https://codecov.io/gh/sourcegraph/sourcegraph/pull/9702?src=pr\u0026el=desc) into [master](https://codecov.io/gh/sourcegraph/sourcegraph/commit/85ce14f49669f51b7761ad752902fa0fed651987\u0026el=desc) will **increase** coverage by `0.00%`.\n\u003e The diff coverage is `75.00%`.\n\n```diff\n@@           Coverage Diff           @@\n##           master    #9702   +/-   ##\n=======================================\n  Coverage   41.99%   41.99%           \n=======================================\n  Files        1339     1339           \n  Lines       73647    73650    +3     \n  Branches     6627     6629    +2     \n=======================================\n+ Hits        30928    30930    +2     \n  Misses      39876    39876           \n- Partials     2843     2844    +1     \n```\n\n| Flag | Coverage Δ | |\n|---|---|---|\n| #unit | `41.99% \u003c75.00%\u003e (+\u003c0.01%)` | :arrow_up: |\n\n| [Impacted Files](https://codecov.io/gh/sourcegraph/sourcegraph/pull/9702?src=pr\u0026el=tree) | Coverage Δ | |\n|---|---|---|\n| [cmd/frontend/graphqlbackend/campaigns.go](https://codecov.io/gh/sourcegraph/sourcegraph/pull/9702/diff?src=pr\u0026el=tree#diff-Y21kL2Zyb250ZW5kL2dyYXBocWxiYWNrZW5kL2NhbXBhaWducy5nbw==) | `0.00% \u003cø\u003e (ø)` | |\n| [...terprise/internal/campaigns/resolvers/campaigns.go](https://codecov.io/gh/sourcegraph/sourcegraph/pull/9702/diff?src=pr\u0026el=tree#diff-ZW50ZXJwcmlzZS9pbnRlcm5hbC9jYW1wYWlnbnMvcmVzb2x2ZXJzL2NhbXBhaWducy5nbw==) | `48.44% \u003c50.00%\u003e (-0.61%)` | :arrow_down: |\n| [...rc/enterprise/campaigns/detail/CampaignDetails.tsx](https://codecov.io/gh/sourcegraph/sourcegraph/pull/9702/diff?src=pr\u0026el=tree#diff-d2ViL3NyYy9lbnRlcnByaXNlL2NhbXBhaWducy9kZXRhaWwvQ2FtcGFpZ25EZXRhaWxzLnRzeA==) | `52.68% \u003c50.00%\u003e (ø)` | |\n| [...nterprise/internal/campaigns/resolvers/resolver.go](https://codecov.io/gh/sourcegraph/sourcegraph/pull/9702/diff?src=pr\u0026el=tree#diff-ZW50ZXJwcmlzZS9pbnRlcm5hbC9jYW1wYWlnbnMvcmVzb2x2ZXJzL3Jlc29sdmVyLmdv) | `35.83% \u003c100.00%\u003e (+0.15%)` | :arrow_up: |\n| [enterprise/internal/campaigns/store.go](https://codecov.io/gh/sourcegraph/sourcegraph/pull/9702/diff?src=pr\u0026el=tree#diff-ZW50ZXJwcmlzZS9pbnRlcm5hbC9jYW1wYWlnbnMvc3RvcmUuZ28=) | `83.91% \u003c100.00%\u003e (ø)` | |\n| [internal/cmd/tracking-issue/main.go](https://codecov.io/gh/sourcegraph/sourcegraph/pull/9702/diff?src=pr\u0026el=tree#diff-aW50ZXJuYWwvY21kL3RyYWNraW5nLWlzc3VlL21haW4uZ28=) | `34.24% \u003c100.00%\u003e (+0.22%)` | :arrow_up: |\n",
            "URL": "https://github.com/sourcegraph/sourcegraph/pull/9702#issuecomment-612883690",
            "CreatedAt": "2020-04-13T12:46:03Z",
            "UpdatedAt": "2020-04-13T12:46:03Z",
            "IncludesCreatedEdit": false
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:merged",
        "key": "eseliger:1586782584000000000",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Actor": {
                "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                "Login": "eseliger",
                "URL": "https://github.com/eseliger"
            },
            "MergeRefName": "master",
            "URL": "https://github.com/sourcegraph/sourcegraph/pull/9702#event-3226982129",
            "Commit": {
                "OID": "9caafbe5f164a2311b81e8403fe8f27654e60a41",
                "Message": "Make campaign.description nullable (#9702)",
                "MessageHeadline": "Make campaign.description nullable (#9702)",
                "URL": "https://github.com/sourcegraph/sourcegraph/commit/9caafbe5f164a2311b81e8403fe8f27654e60a41",
                "Committer": {
                    "AvatarURL": "https://avatars3.githubusercontent.com/u/19864447?v=4",
                    "Email": "noreply@github.com",
                    "Name": "GitHub"
                },
                "CommittedDate": "2020-04-13T12:56:24Z",
                "PushedDate": "2020-04-13T12:56:26Z"
            },
            "CreatedAt": "2020-04-13T12:56:24Z"
        }
    },
    {
        "id": 0,
        "changeset_id": 46,
        "kind": "github:closed",
        "key": "eseliger:1586782585000000000",
        "created_at": "2020-04-29T12:18:15.386252Z",
        "updated_at": "2020-04-29T12:18:15.386252Z",
        "metadata": {
            "Actor": {
                "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
                "Login": "eseliger",
                "URL": "https://github.com/eseliger"
            },
            "CreatedAt": "2020-04-13T12:56:25Z",
            "URL": "https://github.com/sourcegraph/sourcegraph/pull/9702#event-3226982138"
        }
    }
]
```

</details>

(click to expand)

The problem is that there's multiple `github:review_requested` events with the same key (which should be unique per event!):

```json
[
  {
    "id": 0,
    "changeset_id": 46,
    "kind": "github:review_requested",
    "key": "eseliger::1586443388000000000",
    "created_at": "2020-04-29T12:18:15.386252Z",
    "updated_at": "2020-04-29T12:18:15.386252Z",
    "metadata": {
      "Actor": {
        "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
        "Login": "eseliger",
        "URL": "https://github.com/eseliger"
      },
      "RequestedReviewer": {
        "AvatarURL": "",
        "Login": "",
        "URL": ""
      },
      "RequestedTeam": {
        "Name": "",
        "URL": ""
      },
      "CreatedAt": "2020-04-09T14:43:08Z"
    }
  }
  {
    "id": 0,
    "changeset_id": 46,
    "kind": "github:review_requested",
    "key": "eseliger::1586443388000000000",
    "created_at": "2020-04-29T12:18:15.386252Z",
    "updated_at": "2020-04-29T12:18:15.386252Z",
    "metadata": {
      "Actor": {
        "AvatarURL": "https://avatars0.githubusercontent.com/u/19534377?v=4",
        "Login": "eseliger",
        "URL": "https://github.com/eseliger"
      },
      "RequestedReviewer": {
        "AvatarURL": "",
        "Login": "",
        "URL": ""
      },
      "RequestedTeam": {
        "Name": "",
        "URL": ""
      },
      "CreatedAt": "2020-04-09T14:43:08Z"
    }
  }
]
```

They have neither `RequestedReviewer` nor `RequestedTeam`, which is used to build up the unique key:

https://github.com/sourcegraph/sourcegraph/blob/96b8fe4c463447da3575eddfb7f9cb067348bec1/internal/extsvc/github/pulls.go#L343-L351

So I looked at the pull request in question and saw this:

Screen Shot 2020-04-29 at 14 30 58

I'm 99% sure that these reviews were requested from the "campaigns/core" and "campaigns/web" teams, both of which were deleted today (to merge them into @sourcegraph/campaigns).

### PR DESCRIPTION
**Pull request by mrnugget**
This fixes #10264 by deduplicating the events per changeset before
trying to insert them into the database.

Our unique-index on changsetID-kind-key ensures that we can update the
same changeset event if we try to insert it a second time, but it breaks
if we try to insert the same combination in the same query.

Since we shouldn't (see ticket!) have changesetevents with duplicate
Kind+Key+ChangesetID in the first place, it's okay to drop them if we
have them.

### CODE REVIEW

File: enterprise/internal/campaigns/syncer.go [hunk 1]
```diff
@@ -558,7 +558,19 @@ func SyncChangesetsWithSources(ctx context.Context, store SyncStore, bySource []
...
 			SetDerivedState(c.Changeset, csEvents)
 
-			events = append(events, csEvents...)
+			// Deduplicate events per changeset based on their Kind+Key to avoid
+			// conflicts when inserting into database.
+			uniqueEvents := make(map[string]struct{}, len(csEvents))
+			for _, e := range csEvents {
+				k := string(e.Kind) + e.Key
+				if _, ok := uniqueEvents[k]; ok {
+					log15.Info("dropping duplicate changeset event", "changeset_id", e.ChangesetID, "kind", e.Kind, "key", e.Key)
```
Comment (unknwon) [inline enterprise/internal/campaigns/syncer.go:567]: nit
```suggestion
					log15.Debug("dropping duplicate changeset event", "changeset_id", e.ChangesetID, "kind", e.Kind, "key", e.Key)
```

**Comment by unknwon** [comment]
What's wrong with codecov bot, it's not the first time I've seen it posting duplicated comments.

**Comment by eseliger** [comment]
we upload the coverage files at different points in time, so codecov doesn't know for how many reports to wait for :|

File: enterprise/internal/campaigns/syncer.go (hunk 1 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/syncer.go:567]: I think it's more important than debug, actually.

**Pull request merged**